### PR TITLE
feat(task db): implement a task DB

### PR DESCRIPTION
**Pull request by mratsim**
## Overview

When dealing with provers we need to be able to persist state of proving requests in case of network failure or hardware restart or cancellations.

In particular for Bonsai, without storing state, we would be unable to know how many tasks are in the queue, risking waiting in vain if there were none, or overloading the queue by mistake.

## Usage

The DB introduced in this PR should be used either in front of Raiko, from taiko-geth or taiko-reth. Or after raiko. It can be either integrated in an existing product or as a microservice, for example as a daemon or a server.

## Note

This PR currently has no doc as I need to leave it 95% finished due to personal reasons.

For @petarvujovic98, as discussed please review the error model so that it fits Raiko with anyhow, thiserror, ....

I might not be able to update it for several days so feel free to take over.

## API

For now this provides

- A way to enqueue a new task
- A way to update the status(es) of an existing task, including caching a proof and changing proof providers.
- A way to retrieve a cached proof
- A way to get a task status. This returns a vector (provider, status, timestamp) as a task might be in network failure say in Bonsai, and we switch to local proving and it's in-progress.

This will need a way to clear the GuestInput/Payload with "older than Duration" parameter, default 18 days (EIP-4844 blobs expiry) as those are several megabytes.

### CODE REVIEW

**Review by smtmfft — COMMENTED** [review]
The db working model is LGTM. Also V2 is pretty good for client to integrate without any interference with existed system.
However, I am thinking the db design is too heavy for a raiko inside lib, as the general purpose is to record the incoming requests and track the status. If we don't take raiko service reboot into account, a in memory db (even a hashmap) is good enough, i.e., key is incoming request, value can be either proof or status.
current db tables seems more reasonable in big application which multi-services and using db to connect with each other. In the future we maybe need to do that, but that's a long term design.
I think better to use a simpler light weight table and consider these tables as a future extension for independent gateway service (before or after depends on future design, but not inside raiko)

per my understanding, things need to be coverred by db feature in short term:
1. record & tracking task -- will be done by this PR.
2. reboot & recovery the on-going tasks. -- not that critical I guess, if zkVM takes few hours, we will not have long zk queue, if it generates proof very soon, re-run a task is also quick.
3. how to handle reorg request? -- reorg request has the same input but (should have) different result, so maybe client need to inform us this is a re-run instead of query, giving them a 'force' option is fine I think.
4. db prune for long run -- not that critical as Mamy already calculated.

File: host/src/server/api/v2/proof/mod.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            doc.merge(sub_doc);
+            doc
+        })
+}
+
+pub fn create_router() -> Router<ProverState> {
+    Router::new()
+        .merge(get::create_router())
+        .merge(status::create_router())
+        .merge(submit::create_router())
```
[the remark this answers is not in the document]
Comment (petarvujovic98) [reply host/src/server/api/v2/proof/mod.rs:23]: Doesn't this make the API a bit inconsistent? The client than needs to address differently structured responses for a single API route, whereas this way it is clear when we have a status return or a proof return or an task_id return

[the remark this answers is not in the document]
Comment (smtmfft) [reply host/src/server/api/v2/proof/mod.rs:23]: > Doesn't this make the API a bit inconsistent? The client than needs to address differently structured responses for a single API route, whereas this way it is clear when we have a status return or a proof return or an task_id return

I think response could be the same with different status, like {status: running, proof: null} and {status: ok, proof: 0x12345678}. I don't remember how A6 deals with that. Anyway, let's discuss with client team to see which one is perferred.

[the remark this answers is not in the document]
Comment (petarvujovic98) [reply host/src/server/api/v2/proof/mod.rs:23]: Sounds good

**Review by petarvujovic98 — APPROVED** [review]

**Pull request merged**